kong/plugins/ai-proxy/handler.lua

```diff
@@ -10,8 +10,18 @@ local strip = require("kong.tools.utils").strip
 local GCP_SERVICE_ACCOUNT do
   GCP_SERVICE_ACCOUNT = os.getenv("GCP_SERVICE_ACCOUNT")
 end
+local AWS_REGION do
+  AWS_REGION = os.getenv("AWS_REGION")
+end
+local AWS_ACCESS_KEY_ID do
+  AWS_ACCESS_KEY_ID = os.getenv("AWS_ACCESS_KEY_ID")
+end
+local AWS_SECRET_ACCESS_KEY do
+  AWS_SECRET_ACCESS_KEY = os.getenv("AWS_SECRET_ACCESS_KEY")
+end
```

Comment:
```suggestion
local aws_config = require "resty.aws.config"
```

Let `lua-resty-aws` read all the envs it needed, so that other credential chains can be used correctly
The access_key/secret_key/region can also be fetched from it so we don't need to getenv again

Reply:
Hey @windmgc , this is going to sound weird:

so I tried that actually, and the problem is that the AWS SDK cannot read the environment variables AFTER init. In fact they always come out nil.

In this case, I need to instantiate the client at "first access time" (because I need to control some different config / create my own signer class) and so this was the only way to do it.

Reply:
i.e. if you try to create AWS:new() during access phases in any plugin, all env var is nil! This won't work in this case, cos each plugin might have different AWS config.

Reply:
I can see different issues here:
- The plugin's handler is loaded in the `init` phase, and calling `local aws_config = require "resty.aws.config"` on the top of this `handler.lua` file will also let the AWS SDK read env vars in the init phase. So I think it is not happening after `init` phase. Please note that I'm not advising you to initialize a new AWS object here, just letting the SDK read global variables(so that the AWS object can be created using global configurations in the access phase later). This is because that credential chain will need the global variables so you must ensure that those env vars are being read inside `init` phase(and the way is to require `resty.aws.config` but not accessing it)
- Initializing a new AWS object must be delayed during the access phase because as you said each plugin may have different AWS configs, so putting it inside the metastable index function of the per-plugin cache you implemented is the correct way.

Reply:
You can also check `aws-lambda`'s implementation: https://github.com/Kong/kong/blob/40c86fa09710d8e59f01f839eb26e45d27257876/kong/plugins/aws-lambda/handler.lua#L13

All of them are doing env vars fetching in `init` phase, then initializing AWS obj in the later phases(for example `access`)

Reply:
@windmgc Yeah that's right on second one, but first one it just doesn't... work.

If you try to load new AWS handler like:

```
= init phase =

local aws_config = require "resty.aws.config"


= access phase =

local AWS = aws:new()
```

Just fails to authenticate, all env vars are null. Unless I need to pass the "aws_config" object into the :new() constructor?

Reply:
We fixed all of this on Slack, see d3695d193fd8e1019555fa313b2a8c69cc994219